app/src/main/java/it/niedermann/owncloud/notes/util/NoteLinksUtils.java

```diff
@@ -0,0 +1,66 @@
+package it.niedermann.owncloud.notes.util;
+
+import android.text.TextUtils;
+
+import androidx.annotation.VisibleForTesting;
+
+import java.util.HashSet;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class NoteLinksUtils {
+
+    @VisibleForTesting
+    static final String RELATIVE_LINK_WORKAROUND_PREFIX = "https://nextcloudnotes/notes/";
+
+    private static final String linksThatLookLikeNoteLinksRegEx = "\\[[^\\]]*\\]\\((\\d+)\\)";
```

Comment:
This regex looks very complicated. Wouldn't something like `^\[.+\]\(\d+\)$` be easier? To be honest, i do not understand your regex, there are so many escapes :eyes:

Reply:
I agree, that it doesn't look straight forward and an explanatory comment might be needed.

**Some explanation:**

Most of the escapes are not the regex itself, but java string encoding. Java tries to interpret `\` as an escape character itself. So I need to write every `\` as `\\`.

What's left is then `\[[^\]]*\]\((\d+)\)`. Exploded this looks like
```
\[
    [^\]]*           # as much non-`]` characters as possible
\]
\(
    (\d+)            # some integer in a catch group
\)
```
I can't use `.+` in the first group, since this is evaluated greedy. So `[link](123456) and [link](654321)` would be one big match from the first `[` to the very last `]`, instead of two smaller matches.

Reply:
Still open for improvements of course :smiley:

Reply:
Yep, you are right. I'd suggest splitting the regex in the code like you did in your comment above and concatenate the regex, while adding these explanation comments beside the chunks. The compiler will merge the string anyway.

Reply:
I just looked it up, because I was wondering, why the suggested (greedy) regxex even works. In fact, it is greedy, but if the rest of the regex matches, the characters are taken away from the greedy part and the rest matches. So: you can use both expressions.